### PR TITLE
Vis fuse2

### PR DESCRIPTION
**Pull request by itaijj**

### CODE REVIEW

File: fuse/data/ops/ops_visprobe.py [hunk 1]
```diff
@@ -76,111 +126,68 @@ def _extract_collected(self, sample_dict: NDict):
...
-        vis_data = self._extract_data(sample_dict, self._keys, op_id)
-        both_fr =  (VisFlag.REVERSE | VisFlag.FORWARD) in self._flags
-        dir_forward = flow == VisFlag.FORWARD
-        dir_reverse = flow == VisFlag.REVERSE
-        any_show_collected = VisFlag.SHOW_ALL_COLLECTED|VisFlag.SHOW_COLLECTED
-        
-        if VisFlag.COLLECT in self._flags or (dir_forward and both_fr):
+        vis_data = self._extract_data(sample_dict, self._keys ,self._name)
+        name_prefix=""
+        if "name" in sample_dict.to_dict().keys():
```
Comment (mosheraboh) [inline fuse/data/ops/ops_visprobe.py:157]: Can you please remind me what is this name and where we set it?

File: fuse/data/ops/ops_visprobe.py [hunk 2]
```diff
@@ -26,44 +70,50 @@ class VisProbe(OpBase):
...
+    - this Operation is not thread safe and so multithreading is also discouraged
 
     "
     """
 
     def __init__(self,flags: VisFlag, 
                  keys: Union[List, dict] ,
                  type_detector: TypeDetectorBase,
-                 id_filter: Union[None, List] = None, 
+                 coco_converter : Callable = None,
```
Comment (mosheraboh) [inline fuse/data/ops/ops_visprobe.py:81]: coco is too specific,  we might want to use different converters. You can just call it convertor

File: fuse/data/visualizer/visualizer_base.py [hunk 3]
```diff
@@ -1,5 +1,14 @@
...
-from typing import Dict, Any, List
+from typing import Dict, Any, List ,Union , Tuple
+from fuse.utils.ndict import NDict
+from torch import Tensor
+from torchvision.utils import save_image
+import matplotlib
+import numpy as np
+import nibabel as nib
+import os
+from fuseimg.utils.typing.key_types_imaging import DataTypeImaging
```
Comment (mosheraboh) [inline fuse/data/visualizer/visualizer_base.py:10]: fuse.data can't import fuseimg
You shold probably move the imaging part to fuseimg

File: fuse/data/visualizer/visualizer.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import os
+import matplotlib.pyplot as plt
+from typing import List
+from fuseimg.utils.visualization.multi_label import *
```
Comment (mosheraboh) [inline fuse/data/visualizer/visualizer.py:4]: fuse.data can't import fuseimg.
You should move this code to fuseimg.

File: fuse/data/visualizer/visualizer_base.py [hunk 5]
```diff
@@ -46,6 +94,15 @@ def _show(self, vis_data):
...
 
-    def show(self, vis_data):
-        data = self._preprocess(vis_data)
-        self._show(data)
+    def _save(self, vis_data : List , output_path : str , file_name : str ):
+        if self._format == "nii.gz" :
+            for item in vis_data:
+                if isinstance(item.image,Tensor) :
+                    item.image = item.image.numpy()
+                    item.image = nib.Nifti1Image(item.image, affine=None)
```
Comment (mosheraboh) [inline fuse/data/visualizer/visualizer_base.py:102]: I guess this should be outside the if

File: fuseimg/data/ops/image_loader.py [hunk 6]
```diff
@@ -33,4 +34,25 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
         return sample_dict
 
     def reverse(self, sample_dict: NDict, key_to_reverse: str, key_to_follow: str, op_id: Optional[str]) -> dict:        
-        return sample_dict          
+        return sample_dict 
+    
+class OpDownloadImage(OpBase):
```
Comment (mosheraboh) [inline fuseimg/data/ops/image_loader.py:39]: Is it different from OpLoadimage?

File: fuseimg/utils/typing/typed_element.py [hunk 7]
```diff
@@ -32,5 +37,5 @@ def typedElementFromSample(sample_dict, key_pattern, td):
 
 def typedElementToSample(sample_dict, typed_element, keymap):
     for k,v in keymap.items():
-        sample_dict[v] = typed_element.__getattribute__(k.value)
+        NDict.set(sample_dict, v, typed_element.__getattribute__(k.value))
```
Comment (mosheraboh) [inline fuseimg/utils/typing/typed_element.py:40]: why?

File: fuseimg/utils/typing/typed_element.py [hunk 8]
```diff
@@ -2,28 +2,33 @@
...
         self.bboxes = bboxes
+        self.crle = crle
+        self.ucrle = ucrle
         self.labels = labels
         self.metadata = metadata
 
 def typedElementFromSample(sample_dict, key_pattern, td):
     patterns = Patterns({key_pattern: True}, False)
-    all_keys = [k for k in sample_dict.get_all_keys() if patterns.get_value(k)]
+    all_keys = [k for k in NDict.get_all_keys(sample_dict) if patterns.get_value(k)]
```
Comment (mosheraboh) [inline fuseimg/utils/typing/typed_element.py:31]: sample_dict.keypaths() should work

File: fuseimg/utils/visualization/multi_label.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        * cmap[image_index] = value
+            for example - cmap0='gray' will only change the cmap of the first image
+        * unify_size = 'dont_care'
+            will make sure that all images are resized to the same size
+    example usage:
+    imshowmultiple([img1, img2])
+    imshowmultiple(img1,img2,img3, cmap='gray', vmin=0.0, vmax=1.0)
+    imshowmultiple(img1,img2,img3, cmap0='gray')   #will use grayscale color map only on the first image and default cmap on the rest
+    imshowmultiple(img1,img2,img3, unify_size='blah')   #will resize all images to match
+    @param plot_label : function to plot the ground truth segmentation
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/multi_label.py:23]: can you change it to the comment style used in fuse?

File: fuseimg/utils/visualization/tests/inputs/detection/instances_val2014_fakesegm100_results.json [hunk 10]
```diff
@@ -0,0 +1 @@
...ounts":"_in23k>3lNLYC6e<KZC7e<IYC9g<H`BN3O6=U=0_BF<;T=4mBLR=4nBMP=4oBMDCX=`0RC7j<JTC;i<ESC`0k<l0O100T1nNM1OO2O0O1N2N4M6K5J2O1O2N1O2N1OPVn5"},"score":0.994},{"image_id":1270,"category_id":1,"segmentation":{"size":[480,640],"counts":"aZj4>b>7H6K4L2M4M2M2O00100?@`0A`a[4"},"score":0.563},{"image_id":1270,"category_id":1,"segmentation":{"size":[480,640],"counts":"cch35k>5K0O1000O1000ROJRC6m<8[BB47a=>^BBa=o0001O001O00001O001O0010O0001O001L4H>G9I7Hb\\T5"},"score":0.179},{"image_id":1270,"category_id":1,"segmentation":{"size":[480,640],"counts":"eQP16g>5L4M2OO1OHDnA:S>HlA7T>JlA5T>LlA3T>NlA2T>NlA1T>0lAOT>2lAMT>4lALT>4lAKU>5kAJU>7kAIU>7kAIg=3YB50He=8XB03Gd=j0\\BVOd=j0\\BVOc=k0]BTOd=l0\\BROe=?f0I7N20000O10000O2O000O2O1Ocfi7"},"score":0.713},{"image_id":1290,"category_id":1,"segmentation":{"size":[540,640],"counts":"j^n42h`02O1N2O1N3N1N2O1N2O1N2O2M2O1N2N2O1N2O2M2O1N2O1N2O1N3N1N2N2O1N2O1N2O2M2O1N2O1N2O1N3M2O1N2O1N2O1N3M2L4L4K5L4K5L4K6K4L4K5L4K5L4L5J5L4K5L4K5L4L5J5L4K5L4K5L4L4K6K4K5L4L4K5L4K6K4K5L4L21O2N2O1N2N2N2O1N2N2N2N2O1N2N2N2O1N2N2N2O1N2N2N2N2O1N2N2N2O1N2N2N2O1O1N2O1O1O1N2O1O1N2O1O1N2O1O1N2O1O1O0O2O1O1N2O1O1N2O1O1O1N2O1O1N2O1O1N2O1O1N2O1O1O1N2O1O1N2O1O1N2O1O1N2K5I7I7J6I7I7I7G9E;E;E;E;H8K5K5
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/tests/inputs/detection/instances_val2014_fakesegm100_results.json:1]: Can you also put a simple example also in fuse/data/README.md
And a link to the notebook below?

File: fuse/data/ops/tests/test_op_visprobe.py [hunk 11]
```diff
@@ -1,284 +0,0 @@
-import unittest
```
Comment (mosheraboh) [inline fuse/data/ops/tests/test_op_visprobe.py:1]: So where is the test that runs the notebook?

File: fuse/data/visualizer/visualizer_base.py [hunk 12]
```diff
@@ -25,17 +34,56 @@ def _show(self, vis_data: List):
         :param vis_data: preprocessed visualization items to display
         """
         raise "should implement abstract method"
+    
+    @abstractmethod
```
Comment (mosheraboh) [inline fuse/data/visualizer/visualizer_base.py:38]: Why abstract?

File: fuse/data/visualizer/visualizer_base.py [hunk 13]
```diff
@@ -25,17 +34,56 @@ def _show(self, vis_data: List):
...
+        """
+        raise "should implement abstract method"
+    def _preprocess(self, vis_data: Union[List, NDict]) -> Tuple[List[Dict] , str]:
+        if isinstance(vis_data, NDict):
+            vis_data = [vis_data]
+        res = []
+        for element in vis_data :
+            vitem = TypedElement()
+            for item, value in element.items() :
+                if value['type'] is DataTypeImaging.IMAGE:
```
Comment (mosheraboh) [inline fuse/data/visualizer/visualizer_base.py:56]: Where do we set the type and how?

**Review by mosheraboh — CHANGES_REQUESTED** [review]
Thanks a lot Itai.
I guess that it was really hard to follow without Daniel.
Please see my comments inline.

File: fuse/data/ops/ops_visprobe.py (hunk 1 above)
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:157]: it's something I added , you can see in the example.
I wanted to have an option to give specific identifier to each image

File: fuse/data/ops/ops_visprobe.py (hunk 2 above)
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:81]: what I mean is that a user must use COCO format for segmentations or have a function that converts to COCO format.
that's what we talked about before.. 
we don't want to support more than 1 format for each seg type

File: fuse/data/visualizer/visualizer_base.py (hunk 13 above)
Comment (itaijj) [reply fuse/data/visualizer/visualizer_base.py:56]: it's when you give namespace as pipeline key.
example.img
example.seg
etc'

File: fuse/data/visualizer/visualizer.py (hunk 4 above)
Comment (itaijj) [reply fuse/data/visualizer/visualizer.py:4]: sorry, this file should be removed, I accidently copied it here

File: fuse/data/ops/ops_visprobe.py (hunk 1 above)
Comment (mosheraboh) [reply fuse/data/ops/ops_visprobe.py:157]: Can pass the name to OpProbe constructor instead?

File: fuse/data/ops/ops_visprobe.py (hunk 2 above)
Comment (mosheraboh) [reply fuse/data/ops/ops_visprobe.py:81]: It's maybe true for imaging, but about the other domains?
Or if we decide to support more options in the future?

File: fuse/data/visualizer/visualizer_base.py (hunk 13 above)
Comment (mosheraboh) [reply fuse/data/visualizer/visualizer_base.py:56]: Is it a must (the img suffix) if we want to use this visualizer?

File: fuse/data/ops/ops_visprobe.py [hunk 14]
```diff
@@ -76,111 +126,68 @@ def _extract_collected(self, sample_dict: NDict):
...
         res = NDict()
-        for group_id, group_keys in keys.items():
-            for key in group_keys:
-                prekey = f'groups.{group_id}.{key.replace(".", "_")}'
-                res[f'{prekey}.value'] = sample_dict[key]
-                res[f'{prekey}.type'] = self._type_detector.get_type(sample_dict, key)
-        res['$step_id'] = op_id
+        for key in keys:
+            prekey = key.replace(".", "_")
+            if isinstance(sample_dict[key] , Tensor ):
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:133]: done

File: fuse/data/ops/ops_visprobe.py [hunk 15]
```diff
@@ -76,111 +126,68 @@ def _extract_collected(self, sample_dict: NDict):
...
+                res[f'{prekey}.value'] = sample_dict[key].clone()
+            else :
+                 res[f'{prekey}.value'] = sample_dict[key].copy()
+            res[f'{prekey}.type'] = self._type_detector.get_type(sample_dict, key)
+            
+            if self._coco_converter != None :
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:139]: moved it to fuseimg visualizer2D class

File: fuse/data/ops/ops_visprobe.py (hunk 1 above)
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:157]: it already have name there, but it's a step name.
I need also a sample unique name but i'l use id

File: fuse/data/ops/ops_visprobe.py (hunk 2 above)
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:81]: coco converter moved to imaging visualizer

File: fuse/data/visualizer/visualizer_base.py (hunk 3 above)
Comment (itaijj) [reply fuse/data/visualizer/visualizer_base.py:10]: this file is under fuseimg only now

File: fuseimg/utils/visualization/multi_label.py (hunk 9 above)
Comment (itaijj) [reply fuseimg/utils/visualization/multi_label.py:23]: done

File: fuse/data/ops/tests/test_op_visprobe.py (hunk 11 above)
Comment (itaijj) [reply fuse/data/ops/tests/test_op_visprobe.py]: i'l add it

File: fuse/data/visualizer/visualizer_base.py (hunk 12 above)
Comment (itaijj) [reply fuse/data/visualizer/visualizer_base.py:38]: it's part of Daniel's design, I think he wanted common API to all visualizers

File: fuse/data/visualizer/visualizer_base.py (hunk 13 above)
Comment (itaijj) [reply fuse/data/visualizer/visualizer_base.py:56]: I think this is part of the design You did before, 
otherwise why do we need TypedElement?

File: fuseimg/utils/typing/typed_element.py (hunk 8 above)
Comment (itaijj) [reply fuseimg/utils/typing/typed_element.py:31]: removed this unused function

File: fuseimg/utils/visualization/imaging_multi_plot.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
+from typing import List , Callable , Any
+import matplotlib.pyplot as plt
+import numpy as np
+from functools import partial
+from fuseimg.utils.typing.typed_element import TypedElement
+from matplotlib.collections import PatchCollection
+from matplotlib.patches import Polygon
+
+def show_multiple_images(plot_seg : Callable , imgs : List, base_resolution :int = 5, **args ):
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/imaging_multi_plot.py:9]: args to kwargs

File: fuseimg/utils/visualization/imaging_multi_plot.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    '''
+    Show multiple images with shared zoom/translation controls
+    everything passed as kwargs (for example - cmap='gray') will be passed to the individual imshow calls
+    special possible values are:
+        * cmap[image_index] = value
+            for example - cmap0='gray' will only change the cmap of the first image
+        * unify_size = 'dont_care'
+            will make sure that all images are resized to the same size
+    example usage:
+    imshowmultiple([img1, img2])
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/imaging_multi_plot.py:19]: show_multiple_images
Also the first argument is plot_seg

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 16 above)
Comment (mosheraboh) [reply fuseimg/utils/visualization/imaging_multi_plot.py:9]: plot_seg is a must argument?

File: fuseimg/utils/visualization/imaging_multi_plot.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        if 0==i:
+            axis.append(fig.add_subplot(grid_size,grid_size,i+1))
+        else:
+            axis.append(fig.add_subplot(grid_size, grid_size, i + 1, sharex=axis[0],sharey=axis[0]))
+            
+        img_cmap = 'cmap'+str(i+1)
+        if img_cmap in args:
+            pass_kwargs['cmap'] = args[img_cmap]      
+            axis[i].imshow(im, interpolation='none', **pass_kwargs)
+        else:
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/imaging_multi_plot.py:49]: else you use the cmap from the last iteration?

File: fuseimg/utils/visualization/imaging_multi_plot.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    :param args: additional parameters to the plot function of matplotlib
+    :return:
+    '''
+    assert len(imgs)>0
+    grid_size = int(np.sqrt(len(imgs))) + 1
+    fig = plt.figure(figsize=(base_resolution*grid_size,base_resolution*grid_size))
+    axis = []
+    
+    do_not_pass = ['unify_size','color']
+    do_not_pass += [ 'cmap'+str(i) for i in range(20)]
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/imaging_multi_plot.py:35]: Why 20?

File: fuseimg/utils/visualization/imaging_multi_plot.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+from matplotlib.collections import PatchCollection
+from matplotlib.patches import Polygon
+
+def show_multiple_images(plot_seg : Callable , imgs : List, base_resolution :int = 5, **args ):
+    '''
+    Show multiple images with shared zoom/translation controls
+    everything passed as kwargs (for example - cmap='gray') will be passed to the individual imshow calls
+    special possible values are:
+        * cmap[image_index] = value
+            for example - cmap0='gray' will only change the cmap of the first image
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/imaging_multi_plot.py:15]: Do you have a use case for it? 
Maybe get a dict if all the arguments are the same and a sequence of dicts if different arguments per image are required?

File: fuseimg/utils/visualization/imaging_multi_plot.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+        img_cmap = 'cmap'+str(i+1)
+        if img_cmap in args:
+            pass_kwargs['cmap'] = args[img_cmap]      
+            axis[i].imshow(im, interpolation='none', **pass_kwargs)
+        else:
+            axis[i].imshow(im, interpolation='none', **pass_kwargs)
+           
+        plot_seg(axis[i], m )
+                
+        if m.metadata:
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/imaging_multi_plot.py:54]: All the metadata goes into the title - isn't it too much?

File: fuseimg/utils/visualization/imaging_multi_plot.py [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+def plot_color_mask(mask , ax ) :
+    color_mask = np.random.random((1, 3)).tolist()[0]
+    masked = np.ma.masked_where(mask == 0, mask)
+    img = np.ones( (masked.shape[0], masked.shape[1], 3) )
+    for i in range(3):
+        img[:,:,i] = color_mask[i]
+    ax.imshow(np.dstack( (img, masked*0.5) )) 
+    
+    
+def plot_seg_coco_style(ax : Any, sample : TypedElement ):
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/imaging_multi_plot.py:70]: What is coco style?

File: fuseimg/utils/visualization/imaging_multi_plot.py [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+        plot_seg(axis[i], m )
+                
+        if m.metadata:
+            axis[i].set_title(str(i)+":"+m.metadata)
+        else:
+            axis[i].set_title(str(i))
+
+    return fig
+
+def plot_color_mask(mask , ax ) :
```
Comment (mosheraboh) [inline fuseimg/utils/visualization/imaging_multi_plot.py:61]: annotation type and documentation

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 16 above)
Comment (mosheraboh) [reply fuseimg/utils/visualization/imaging_multi_plot.py:9]: return value type is missing.

**Review by mosheraboh — CHANGES_REQUESTED** [review]
Thanks Itai!
Please find my comments inline

File: fuse/data/README.md [hunk 24]
```diff
@@ -61,7 +61,73 @@ A pipeline is created from a list of tuples. Each tuple includes an op and op ar
 In this example "sample_id" is a running index. OpKits21SampleIDDecode() is a custom op for Kits21 challenge converting the index to image path and segmentation path which are then loaded by OpLoadImage().
 In other case than Kits21 you would have to implement your custome MySampleIDDecode() operator.
 Finally, OpClip() and OpToRange() pre-process the image.
- 
+
+## Visualization - VisProbOp
+This op can be placed between different steps at the pipeline , it must be used only in the pipeline.
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/README.md:66]: ok, I can remove it as every op must be used in pipeline.
register the image to be visualized inline/later

File: fuse/data/README.md [hunk 25]
```diff
@@ -61,7 +61,73 @@ A pipeline is created from a list of tuples. Each tuple includes an op and op ar
 In this example "sample_id" is a running index. OpKits21SampleIDDecode() is a custom op for Kits21 challenge converting the index to image path and segmentation path which are then loaded by OpLoadImage().
 In other case than Kits21 you would have to implement your custome MySampleIDDecode() operator.
 Finally, OpClip() and OpToRange() pre-process the image.
- 
+
+## Visualization - VisProbOp
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/README.md:65]: here I just talk in general on the operation, it's not specific for 2D images.

File: fuse/data/README.md [hunk 26]
```diff
@@ -61,7 +61,73 @@ A pipeline is created from a list of tuples. Each tuple includes an op and op ar
...
+This op can be placed between different steps at the pipeline , it must be used only in the pipeline.
+
+Each step can have an optional name which will appear in every image and file created by the visualizer.
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/README.md]: op

File: fuse/data/README.md [hunk 27]
```diff
@@ -61,7 +61,73 @@ A pipeline is created from a list of tuples. Each tuple includes an op and op ar
...
+
+## Basic example - using Imaging2dVisualizer
+in each step we call VisProbe with same flag VisFlag.COLLECT and in the end we call with VisFlag.SHOW_COLLECTED
+
+because we are using ONLINE mode images will be displayed in console.
+
+The follwing example donwloads kits21 image and segmentation and visualize it's rotation and in a specific slice ( so we can use 2d visualizer)
+
+**The original code is in fuseimg/utils/visualization/visualization_example.ipynb**
+```python
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/README.md:96]: I can have less calls to OpVisProb,
we need medical image with segmentation I think, 
you have other example with ready pipeline?

File: fuse/data/ops/ops_visprobe.py [hunk 28]
```diff
@@ -1,69 +1,72 @@
...
-import copy
+from typing import List, Optional, Union , Dict , Any , Callable
 import enum
+import numpy as np
+from copy import deepcopy
 
 from fuse.utils.ndict import NDict
 from fuse.data.visualizer.visualizer_base import VisualizerBase
-from .op_base import OpBase
+from fuseimg.utils.typing.key_types_imaging import DataTypeImaging
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:8]: I did not use, just forgot to remove the import

File: fuse/data/ops/ops_visprobe.py [hunk 29]
```diff
@@ -1,69 +1,72 @@
...
-              OFFLINE - show operations will write to disk (using the caching mechanism) the relevant info (state or states for comparison)                
+              ONLINE - visualize operations will prompt the user with the releveant plot                                                                        
+              OFFLINE - visualize operations will write to disk (using the caching mechanism) the relevant info (state or states for comparison)                
               FORWARD - visualization operation will be activated on forward pipeline execution flow                                                       
               REVERSE - visualization operation will be activated on reverse pipeline execution flow
           :param keys: for which sample keys to handle visualization, also can be grouped in a dictionary
-          :param id_filter: for which sample id's to be activated, if None, active for all samples  
+          :param type_detector : class used to identify objects from the keys
+          :param coco_converter : function that convert the input format to COCO type format
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:54]: Sorry, I just forgot to remove it in the param documentation

File: fuse/data/visualizer/visualizer_base.py [hunk 30]
```diff
@@ -1,51 +1,53 @@
...
+        actual visualization function, gets a preprocessed collection of items to visualize/compare and shows
+        a visualization window that is blocking.
+        should be overriden by a specific visualizer
         
-        :param vis_data: the collected data
+        :param vis_data: preprocessed visualization items to display
         """
-        return vis_data
-
+        raise "should implement abstract method"
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/visualizer/visualizer_base.py:23]: replaced

File: fuse/data/visualizer/visualizer_base.py [hunk 31]
```diff
@@ -1,51 +1,53 @@
...
-            print(f"comparing {len(vis_data)} items:")
-            for item in vis_data:
-                print(item)
-
-    def show(self, vis_data):
-        data = self._preprocess(vis_data)
-        self._show(data)
+        :param vis_data: the collected data
+        """
+        raise "should implement abstract method"
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/visualizer/visualizer_base.py:45]: I think it was the original design , if it is optional than it should be removed? 
This is just an API here

File: fuseimg/data/ops/image_loader.py [hunk 32]
```diff
@@ -35,3 +36,23 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
...
         return sample_dict
 
+class OpDownloadImage(OpReversibleBase):
+    '''
+    Loads a medical image, currently only nii is supported
+    '''
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out: str, format:str="infer"):
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:46]: not sure what was the original meaning

[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:46]: removed

File: fuseimg/data/ops/image_loader.py [hunk 33]
```diff
@@ -35,3 +36,23 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key_in:str, key_out
     def reverse(self, sample_dict: dict, key_to_reverse: str, key_to_follow: str, op_id: Optional[str]) -> dict:
         return sample_dict
 
+class OpDownloadImage(OpReversibleBase):
+    '''
+    Loads a medical image, currently only nii is supported
+    '''
+    def __init__(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/data/ops/image_loader.py:43]: removed

File: fuseimg/utils/visualization/tests/test_visualization.py [hunk 34]
```diff
@@ -0,0 +1,54 @@
+import os
+import pathlib
+import shutil
+import unittest
+from tqdm import trange
+from testbook import testbook
+
+notebook_path = os.path.join(pathlib.Path(__file__).parent.resolve(), "../visualization_example.ipynb")
+
+class TestDatasets(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/utils/visualization/tests/test_visualization.py:10]: changed

File: fuseimg/utils/visualization/tests/test_visualization.py [hunk 35]
```diff
@@ -0,0 +1,54 @@
+import os
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/utils/visualization/tests/test_visualization.py:1]: agreed and changed

File: fuseimg/utils/visualization/save_to_format.py [hunk 36]
```diff
@@ -0,0 +1,125 @@
...
+    :param segmentation_type: DataTypeImaging
+    :return  output mask 
+    """     
+    if segmentation_type == DataTypeImaging.UCRLE:
+        elements = [convert_uncompressed_RLE_COCO_type(element,height,width)  for element in elements]    
+    elif segmentation_type == DataTypeImaging.CRLE:
+        elements = [convert_compressed_RLE_COCO_type(element,height,width)  for element in elements]     
+    return elements
+
+class VisualizerImaging(VisualizerBase):
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/utils/visualization/save_to_format.py:57]: it's just an abstract class that I did, to implement pre_process for images to be common

File: fuseimg/utils/visualization/tests/test_visualization.py [hunk 37]
```diff
@@ -0,0 +1,54 @@
...
+
+        # tb.inject(
+        #     """
+        #     assert(my_dataset[0]["data.gt.seg"].shape[1:] == (4, 256, 256))
+        #     """
+        # )
+
+    
+    def tearDown(self) -> None:
+        # shutil.rmtree(self.kits21_cache_dir)
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/utils/visualization/tests/test_visualization.py:46]: ok

File: fuseimg/utils/visualization/visualizer2D.py [hunk 38]
```diff
@@ -0,0 +1,30 @@
...
+from typing import List
+from fuseimg.utils.visualization.imaging_multi_plot import *
+from typing import Callable, Dict, Any, List ,Union , Tuple
+from fuse.utils.ndict import NDict
+from fuseimg.utils.visualization.save_to_format import VisualizerImaging
+
+        
+class Imaging2dVisualizer(VisualizerImaging):
+    """
+    Curently supports only one group per viewed item
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/utils/visualization/visualizer2D.py:12]: ok

File: fuse/data/ops/ops_visprobe.py [hunk 39]
```diff
@@ -1,69 +1,72 @@
...
-    COLLECT = 1       #save current state for future comparison        
-    SHOW_CURRENT = 2          #show current state                                    
-    SHOW_COLLECTED = 4 #show comparison of all previuosly collected states
-    CLEAR = 8           #clear all collected states until this point in the pipeline
-    ONLINE = 16         #show operations will prompt the user with the releveant plot 
-    OFFLINE = 32        #show operations will write to disk (using the caching mechanism) the relevant info (state or states for comparison) 
-    FORWARD = 64        #visualization operation will be activated on forward pipeline execution flow
-    REVERSE = 128       #visualization operation will be activated on reverse pipeline execution flow
-    SHOW_ALL_COLLECTED = 256 #show comparison of all previuosly collected states
+    COLLECT = 1                     #save current state for future comparison
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:14]: I don't think we have same meaning for comparison ( btw Daniel wrote it )
here it mean you show it as a collection of images, the comparison is always manual.

File: fuseimg/utils/visualization/visualizer2D.py [hunk 40]
```diff
@@ -0,0 +1,30 @@
...
+    """
+    def __init__(self, cmap :str, format : str ="png", coco_converter : Callable = None) -> None:
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuseimg/utils/visualization/visualizer2D.py:14]: ok

File: fuse/data/ops/ops_visprobe.py [hunk 41]
```diff
@@ -1,69 +1,72 @@
...
     - running in a cached environment is dangerous and is prohibited
-    - this Operation is not thread safe ans so multithreading is also discouraged
+    - this Operation is not thread safe and so multithreading is also discouraged
 
     "
     """
 
     def __init__(self,flags: VisFlag, 
                  keys: Union[List, dict] ,
                  type_detector: TypeDetectorBase,
```
[the remark this answers is not in the document]
Comment (itaijj) [reply fuse/data/ops/ops_visprobe.py:38]: I think VisOp should not be used with dataset at all,
only when you call the pipeline on specific example, so we don't have multithreading issues

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 16 above)
Comment (itaijj) [reply fuseimg/utils/visualization/imaging_multi_plot.py:9]: changed.
it's a must, I think the idea that you can replace how you draw the segmentation

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 17 above)
Comment (itaijj) [reply fuseimg/utils/visualization/imaging_multi_plot.py:19]: updated

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 18 above)
Comment (itaijj) [reply fuseimg/utils/visualization/imaging_multi_plot.py:49]: yes

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 19 above)
Comment (itaijj) [reply fuseimg/utils/visualization/imaging_multi_plot.py:35]: removed the list of cmaps

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 20 above)
Comment (itaijj) [reply fuseimg/utils/visualization/imaging_multi_plot.py:15]: removed

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 21 above)
Comment (itaijj) [reply fuseimg/utils/visualization/imaging_multi_plot.py:54]: it's just the name of parameter, in current visualizer implementation it passes the name.
other visualizers may change it

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 22 above)
Comment (itaijj) [reply fuseimg/utils/visualization/imaging_multi_plot.py:70]: displayed COCO types as in COCO library

File: fuseimg/utils/visualization/imaging_multi_plot.py (hunk 23 above)
Comment (itaijj) [reply fuseimg/utils/visualization/imaging_multi_plot.py:61]: removed
